src/reflect/scala/reflect/internal/transform/Erasure.scala

```diff
@@ -388,6 +446,191 @@ trait Erasure {
+     *  This ensure that `A <= B && B <= A` iff `A =:= B`.
+     */
+    def compareErasedGlb(tp1: Type, tp2: Type): Int = {
+      // this check is purely an optimization.
+      if (tp1 eq tp2) return 0
+
+      val isEVT1 = tp1.isInstanceOf[ErasedValueType]
+      val isEVT2 = tp2.isInstanceOf[ErasedValueType]
+      if (isEVT1 && isEVT2) {
+        return compareErasedGlb(
+          tp1.asInstanceOf[ErasedValueType].valueClazz.tpe_*,
+          tp2.asInstanceOf[ErasedValueType].valueClazz.tpe_*)
+      }
+      else if (isEVT1)
+        return -1
+      else if (isEVT2)
+        return 1
+
+      val sym1 = tp1.baseClasses.head
+      val sym2 = tp2.baseClasses.head
+
+      def compareClasses: Int = {
+        if (sym1.isSubClass(sym2))
+          -1
+        else if (sym2.isSubClass(sym1))
+          1
+        else
+          sym1.fullName.compareTo(sym2.fullName)
+      }
+
+      val isArray1 = tp1.typeArgs.nonEmpty && sym1.isSubClass(definitions.ArrayClass)
+      val isArray2 = tp2.typeArgs.nonEmpty && sym2.isSubClass(definitions.ArrayClass)
+      if (isArray1 && isArray2)
+        return compareErasedGlb(tp1.typeArgs.head, tp2.typeArgs.head)
+      else if (isArray1)
+        return -1
+      else if (isArray2)
+        return 1
+
+      val isPrimitive1 = sym1.isPrimitiveValueClass
+      val isPrimitive2 = sym2.isPrimitiveValueClass
+      if (isPrimitive1 && isPrimitive2)
+        return compareClasses
+      else if (isPrimitive1)
+        return -1
+      else if (isPrimitive2)
+        return 1
+
+      val isRealClass1 = sym1.isClass && !sym1.isTrait
+      val isRealClass2 = sym2.isClass && !sym2.isTrait
+      if (isRealClass1 && isRealClass2)
+        return compareClasses
+      else if (isRealClass1)
+        return -1
+      else if (isRealClass2)
+        return 1
+
+      compareClasses
+    }
+
+    components.min((t, u) => compareErasedGlb(t, u))
+  }
+
+  /** For a type alias, get its info as seen from
+   *  the current prefix and owner.
+   *  Sees through opaque type aliases.
+   */
+  def transparentDealias(sym: Symbol, pre: Type, owner: Symbol) = {
+    @inline def visible(tp: Type) = tp.asSeenFrom(pre, owner)
+
+    if (sym.isScala3Defined && !sym.isClass)
+      sym.attachments.get[DottyOpaqueTypeAlias]
+        .map(alias => visible(alias.tpe))
+        .getOrElse(visible(sym.info))
+    else
+      visible(sym.info)
+  }
+
+  /** Dotty implementation of Array Erasure:
+   *
+   *  Is `Array[tp]` a generic Array that needs to be erased to `Object`?
+   *  This is true if among the subtypes of `Array[tp]` there is either:
+   *  - both a reference array type and a primitive array type
+   *    (e.g. `Array[_ <: Int | String]`, `Array[_ <: Any]`)
+   *  - or two different primitive array types (e.g. `Array[_ <: Int | Double]`)
+   *  In both cases the erased lub of those array types on the JVM is `Object`.
+   */
+  def isGenericArrayElement(tp: Type): Boolean = {
```

Comment:
could also be moved to `Scala3ErasureMap`

Reply:
done